### ISSUE TITLE
Typescript type errors in vue-flicking

### ISSUE THREAD

**Issue by justvidyadhar** [naver/egjs-flicking#453]
## Description
I recently setup vue-flicking for carousels in a new Vue project I'm working on(in typescript). But when I build it, it fails with a bunch of errors that point to node_modules. Not sure if this is an error from my end or from the package.

```
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(29,13):
29:13 Property '_el' has no initializer and is not definitely assigned in the constructor.
    27 |   // Internal states
    28 |   protected _flicking: Flicking | null;
  > 29 |   protected _el: HTMLElement;
       |             ^
    30 |   protected _transform: string;
    31 |   protected _position: number;
    32 |   protected _alignPos: number;
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(30,13):
30:13 Property '_transform' has no initializer and is not definitely assigned in the constructor.
    28 |   protected _flicking: Flicking | null;
    29 |   protected _el: HTMLElement;
  > 30 |   protected _transform: string;
       |             ^
    31 |   protected _position: number;
    32 |   protected _alignPos: number;
    33 |   protected _offset: number;
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(31,13):
31:13 Property '_position' has no initializer and is not definitely assigned in the constructor.
    29 |   protected _el: HTMLElement;
    30 |   protected _transform: string;
  > 31 |   protected _position: number;
       |             ^
    32 |   protected _alignPos: number;
    33 |   protected _offset: number;
    34 |   protected _range: { min: number; max: number };
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(32,13):
32:13 Property '_alignPos' has no initializer and is not definitely assigned in the constructor.
    30 |   protected _transform: string;
    31 |   protected _position: number;
  > 32 |   protected _alignPos: number;
       |             ^
    33 |   protected _offset: number;
    34 |   protected _range: { min: number; max: number };
    35 |   protected _visiblePanels: Panel[];
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(33,13):
33:13 Property '_offset' has no initializer and is not definitely assigned in the constructor.
    31 |   protected _position: number;
    32 |   protected _alignPos: number;
  > 33 |   protected _offset: number;
       |             ^
    34 |   protected _range: { min: number; max: number };
    35 |   protected _visiblePanels: Panel[];
    36 |   protected _anchors: AnchorPoint[];
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(34,13):
34:13 Property '_range' has no initializer and is not definitely assigned in the constructor.
    32 |   protected _alignPos: number;
    33 |   protected _offset: number;
  > 34 |   protected _range: { min: number; max: number };
       |             ^
    35 |   protected _visiblePanels: Panel[];
    36 |   protected _anchors: AnchorPoint[];
    37 |   protected _needPanelTriggered: { prev: boolean; next: boolean };
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(35,13):
35:13 Property '_visiblePanels' has no initializer and is not definitely assigned in the constructor.
    33 |   protected _offset: number;
    34 |   protected _range: { min: number; max: number };
  > 35 |   protected _visiblePanels: Panel[];
       |             ^
    36 |   protected _anchors: AnchorPoint[];
    37 |   protected _needPanelTriggered: { prev: boolean; next: boolean };
    38 |
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(36,13):
36:13 Property '_anchors' has no initializer and is not definitely assigned in the constructor.
    34 |   protected _range: { min: number; max: number };
    35 |   protected _visiblePanels: Panel[];
  > 36 |   protected _anchors: AnchorPoint[];
       |             ^
    37 |   protected _needPanelTriggered: { prev: boolean; next: boolean };
    38 |
    39 |   // Internal states getter
ERROR in ~/myproject/node_modules/@egjs/flicking/src/camera/Camera.ts(37,13):
37:13 Property '_needPanelTriggered' has no initializer and is not definitely assigned in the constructor.
    35 |   protected _visiblePanels: Panel[];
    36 |   protected _anchors: AnchorPoint[];
  > 37 |   protected _needPanelTriggered: { prev: boolean; next: boolean };
       |             ^
    38 |
    39 |   // Internal states getter
    40 |   /**
ERROR in ~/myproject/node_modules/@egjs/flicking/src/cfc/withFlickingMethods.ts(21,14):
21:14 Argument of type '(name: keyof Flicking) => void' is not assignable to parameter of type '(value: string, index: number, array: string[]) => void'.
  Types of parameters 'name' and 'value' are incompatible.
    Type 'string' is not assignable to type '"resize" | "prev" | "next" | "align" | "horizontal" | "index" | "duration" | "deceleration" | "interruptable" | "easing" | "inputType" | "iOSEdgeSwipeThreshold" | "bounce" | ... 45 more ... | "off"'.
    19 |   Object.getOwnPropertyNames(Flicking.prototype)
    20 |     .filter(name => !prototype[name] && !name.startsWith("_") && name !== "constructor")
  > 21 |     .forEach((name: keyof Flicking) => {
       |              ^
    22 |       const descriptor = Object.getOwnPropertyDescriptor(Flicking.prototype, name)!;
    23 |
    24 |       if (descriptor.value) {
ERROR in ~/myproject/node_modules/@egjs/flicking/src/control/AxesController.ts(21,11):
21:11 Property '_flicking' has no initializer and is not definitely assigned in the constructor.
    19 |  */
    20 | class AxesController {
  > 21 |   private _flicking: Flicking | null;
       |           ^
    22 |   private _axes: Axes | null;
    23 |   private _panInput: PanInput | null;
    24 |   private _stateMachine: StateMachine;
ERROR in ~/myproject/node_modules/@egjs/flicking/src/control/AxesController.ts(22,11):
22:11 Property '_axes' has no initializer and is not definitely assigned in the constructor.
    20 | class AxesController {
    21 |   private _flicking: Flicking | null;
  > 22 |   private _axes: Axes | null;
       |           ^
    23 |   private _panInput: PanInput | null;
    24 |   private _stateMachine: StateMachine;
    25 |
ERROR in ~/myproject/node_modules/@egjs/flicking/src/control/AxesController.ts(23,11):
23:11 Property '_panInput' has no initializer and is not definitely assigned in the constructor.
    21 |   private _flicking: Flicking | null;
    22 |   private _axes: Axes | null;
  > 23 |   private _panInput: PanInput | null;
       |           ^
    24 |   private _stateMachine: StateMachine;
    25 |
    26 |   private _animatingContext: { start: number; end: number; offset: number };
ERROR in ~/myproject/node_modules/@egjs/flicking/src/control/AxesController.ts(26,11):
26:11 Property '_animatingContext' has no initializer and is not definitely assigned in the constructor.
    24 |   private _stateMachine: StateMachine;
    25 |
  > 26 |   private _animatingContext: { start: number; end: number; offset: number };
       |           ^
    27 |
    28 |   /**
    29 |    * An {@link https://naver.github.io/egjs-axes/release/latest/doc/eg.Axes.html Axes} instance
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(20,13):
20:13 Property '_pos' has no initializer and is not definitely assigned in the constructor.
    18 |   protected _flicking: Flicking;
    19 |   protected _index: number;
  > 20 |   protected _pos: number;
       |             ^
    21 |   protected _size: number;
    22 |   protected _height: number;
    23 |   protected _margin: { prev: number; next: number };
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(21,13):
21:13 Property '_size' has no initializer and is not definitely assigned in the constructor.
    19 |   protected _index: number;
    20 |   protected _pos: number;
  > 21 |   protected _size: number;
       |             ^
    22 |   protected _height: number;
    23 |   protected _margin: { prev: number; next: number };
    24 |   protected _alignPos: number; // Actual align pos
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(22,13):
22:13 Property '_height' has no initializer and is not definitely assigned in the constructor.
    20 |   protected _pos: number;
    21 |   protected _size: number;
  > 22 |   protected _height: number;
       |             ^
    23 |   protected _margin: { prev: number; next: number };
    24 |   protected _alignPos: number; // Actual align pos
    25 |   protected _removed: boolean;
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(23,13):
23:13 Property '_margin' has no initializer and is not definitely assigned in the constructor.
    21 |   protected _size: number;
    22 |   protected _height: number;
  > 23 |   protected _margin: { prev: number; next: number };
       |             ^
    24 |   protected _alignPos: number; // Actual align pos
    25 |   protected _removed: boolean;
    26 |   protected _toggleDirection: ValueOf<typeof DIRECTION>;
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(24,13):
24:13 Property '_alignPos' has no initializer and is not definitely assigned in the constructor.
    22 |   protected _height: number;
    23 |   protected _margin: { prev: number; next: number };
  > 24 |   protected _alignPos: number; // Actual align pos
       |             ^
    25 |   protected _removed: boolean;
    26 |   protected _toggleDirection: ValueOf<typeof DIRECTION>;
    27 |   protected _toggled: boolean;
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(26,13):
26:13 Property '_toggleDirection' has no initializer and is not definitely assigned in the constructor.
    24 |   protected _alignPos: number; // Actual align pos
    25 |   protected _removed: boolean;
  > 26 |   protected _toggleDirection: ValueOf<typeof DIRECTION>;
       |             ^
    27 |   protected _toggled: boolean;
    28 |   protected _togglePosition: number;
    29 |
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(27,13):
27:13 Property '_toggled' has no initializer and is not definitely assigned in the constructor.
    25 |   protected _removed: boolean;
    26 |   protected _toggleDirection: ValueOf<typeof DIRECTION>;
  > 27 |   protected _toggled: boolean;
       |             ^
    28 |   protected _togglePosition: number;
    29 |
    30 |   // Options
ERROR in ~/myproject/node_modules/@egjs/flicking/src/core/panel/Panel.ts(28,13):
28:13 Property '_togglePosition' has no initializer and is not definitely assigned in the constructor.
    26 |   protected _toggleDirection: ValueOf<typeof DIRECTION>;
    27 |   protected _toggled: boolean;
  > 28 |   protected _togglePosition: number;
       |             ^
    29 |
    30 |   // Options
    31 |   protected _align: PanelOptions["align"];
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(161,44):
161:44 No overload matches this call.
  Overload 1 of 3, '(callbackfn: (previousValue: { index: number; elements: any[]; }, currentValue: { index: number; elements: any[]; }, currentIndex: number, array: { index: number; elements: any[]; }[]) => { index: number; elements: any[]; }, initialValue: { ...; }): { ...; }', gave the following error.
    Argument of type '(addedPanels: never[], item: { index: number; elements: any[]; }) => Panel[]' is not assignable to parameter of type '(previousValue: { index: number; elements: any[]; }, currentValue: { index: number; elements: any[]; }, currentIndex: number, array: { index: number; elements: any[]; }[]) => { index: number; elements: any[]; }'.
      Types of parameters 'addedPanels' and 'previousValue' are incompatible.
        Type '{ index: number; elements: any[]; }' is missing the following properties from type 'never[]': length, pop, push, concat, and 28 more.
  Overload 2 of 3, '(callbackfn: (previousValue: never[], currentValue: { index: number; elements: any[]; }, currentIndex: number, array: { index: number; elements: any[]; }[]) => never[], initialValue: never[]): never[]', gave the following error.
    Argument of type '(addedPanels: never[], item: { index: number; elements: any[]; }) => Panel[]' is not assignable to parameter of type '(previousValue: never[], currentValue: { index: number; elements: any[]; }, currentIndex: number, array: { index: number; elements: any[]; }[]) => never[]'.
      Type 'Panel[]' is not assignable to type 'never[]'.
        Type 'Panel' is not assignable to type 'never'.
    159 |     const align = this._getPanelAlign();
    160 |
  > 161 |     const allPanelsInserted = items.reduce((addedPanels, item) => {
        |                                            ^
    162 |       const insertingIdx = getMinusCompensatedIndex(item.index, panels.length);
    163 |       const panelsPushed = panels.slice(insertingIdx);
    164 |       const panelsInserted = item.elements.map(el => this._createPanel(el, { index: insertingIdx, align, flicking }));
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(185,27):
185:27 Property 'length' does not exist on type '{ index: number; elements: any[]; }'.
    183 |     }, []);
    184 |
  > 185 |     if (allPanelsInserted.length <= 0) return [];
        |                           ^
    186 |
    187 |     // Update camera & control
    188 |     this._updateCameraAndControl();
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(194,27):
194:27 Property 'length' does not exist on type '{ index: number; elements: any[]; }'.
    192 |     // Move to the first panel added if no panels existed
    193 |     // FIXME: fix for animating case
  > 194 |     if (allPanelsInserted.length > 0 && !control.animating) {
        |                           ^
    195 |       void control.moveToPanel(control.activePanel || allPanelsInserted[0], {
    196 |         duration: 0
    197 |       }).catch(() => void 0);
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(202,5):
202:5 Type '{ index: number; elements: any[]; }' is missing the following properties from type 'Panel[]': length, pop, push, concat, and 28 more.
    200 |     flicking.camera.updateOffset();
    201 |
  > 202 |     return allPanelsInserted;
        |     ^
    203 |   }
    204 |
    205 |   /**
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(222,43):
222:43 No overload matches this call.
  Overload 1 of 3, '(callbackfn: (previousValue: { index: number; deleteCount: number; }, currentValue: { index: number; deleteCount: number; }, currentIndex: number, array: { index: number; deleteCount: number; }[]) => { index: number; deleteCount: number; }, initialValue: { ...; }): { ...; }', gave the following error.
    Argument of type '(removed: never[], item: { index: number; deleteCount: number; }) => Panel[]' is not assignable to parameter of type '(previousValue: { index: number; deleteCount: number; }, currentValue: { index: number; deleteCount: number; }, currentIndex: number, array: { index: number; deleteCount: number; }[]) => { index: number; deleteCount: number; }'.
      Types of parameters 'removed' and 'previousValue' are incompatible.
        Type '{ index: number; deleteCount: number; }' is missing the following properties from type 'never[]': length, pop, push, concat, and 28 more.
  Overload 2 of 3, '(callbackfn: (previousValue: never[], currentValue: { index: number; deleteCount: number; }, currentIndex: number, array: { index: number; deleteCount: number; }[]) => never[], initialValue: never[]): never[]', gave the following error.
    Argument of type '(removed: never[], item: { index: number; deleteCount: number; }) => Panel[]' is not assignable to parameter of type '(previousValue: never[], currentValue: { index: number; deleteCount: number; }, currentIndex: number, array: { index: number; deleteCount: number; }[]) => never[]'.
      Type 'Panel[]' is not assignable to type 'never[]'.
    220 |     const activeIndex = control.activeIndex;
    221 |
  > 222 |     const allPanelsRemoved = items.reduce((removed, item) => {
        |                                           ^
    223 |       const { index, deleteCount } = item;
    224 |       const removingIdx = getMinusCompensatedIndex(index, panels.length);
    225 |
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(256,26):
256:26 Property 'length' does not exist on type '{ index: number; deleteCount: number; }'.
    254 |
    255 |     // FIXME: fix for animating case
  > 256 |     if (allPanelsRemoved.length > 0 && !control.animating) {
        |                          ^
    257 |       const targetPanel = includes(allPanelsRemoved, activePanel)
    258 |         ? (panels[activeIndex] || panels[panels.length - 1])
    259 |         : activePanel;
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(257,36):
257:36 Argument of type '{ index: number; deleteCount: number; }' is not assignable to parameter of type 'unknown[]'.
  Type '{ index: number; deleteCount: number; }' is missing the following properties from type 'unknown[]': length, pop, push, concat, and 28 more.
    255 |     // FIXME: fix for animating case
    256 |     if (allPanelsRemoved.length > 0 && !control.animating) {
  > 257 |       const targetPanel = includes(allPanelsRemoved, activePanel)
        |                                    ^
    258 |         ? (panels[activeIndex] || panels[panels.length - 1])
    259 |         : activePanel;
    260 |
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/Renderer.ts(273,5):
273:5 Type '{ index: number; deleteCount: number; }' is missing the following properties from type 'Panel[]': length, pop, push, concat, and 28 more.
    271 |     flicking.camera.updateOffset();
    272 |
  > 273 |     return allPanelsRemoved;
        |     ^
    274 |   }
    275 |
    276 |   protected _getPanelAlign() {
ERROR in ~/myproject/node_modules/@egjs/flicking/src/renderer/VanillaRenderer.ts(60,7):
60:7 Argument of type '(el: HTMLElement, index: number) => ElementPanel' is not assignable to parameter of type '(value: Element, index: number, array: Element[]) => ElementPanel'.
  Types of parameters 'el' and 'value' are incompatible.
    Type 'Element' is missing the following properties from type 'HTMLElement': accessKey, accessKeyLabel, autocapitalize, dir, and 106 more.
    58 |
    59 |     this._panels = cameraChilds.map(
  > 60 |       (el: HTMLElement, index: number) => new ElementPanel({ flicking, el, index, align })
       |       ^
    61 |     );
    62 |   }
    63 |
```

## Steps to check or reproduce
@vue/cli 4.5.13
Typescript Version: 3.9.7

**Comment by WoodNeck** [comment]
Hello @justvidyadhar!
Thanks for reporting the issue, I found out that it's related to values export in our build script.
I'll leave a comment here after releasing a new version that fix this issue,
While then, you can disable that issue by setting `strict` to false in your tsconfig.json

**Comment by justvidyadhar** [comment]
@WoodNeck Thanks for the swift response.

**Comment by WoodNeck** [comment]
@justvidyadhar I've just released vue-flicking's version 4.0.2 that fixes this issue.
```sh
npm i @egjs/vue-flicking@4.0.2
```
Please test it and leave a comment here if there're more issues :)
